bpf/process/bpf_rate.h

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+/* Copyright Authors of Cilium */
+
+#ifndef __RATE_H__
+#define __RATE_H__
+
+#include "bpf_tracing.h"
+#include "bpf_helpers.h"
+
+struct cgroup_rate_key {
+	__u64 id;
+};
+
+struct cgroup_rate_value {
+	__u64 curr;
+	__u64 prev;
+	__u64 time;
+	__u64 rate;
+	__u64 throttled;
+};
+
+struct cgroup_rate_options {
+	__u64 events;
+	__u64 interval;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+	__uint(max_entries, 32768);
+	__type(key, struct cgroup_rate_key);
+	__type(value, struct cgroup_rate_value);
+} cgroup_rate_map SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, __u32);
+	__type(value, struct cgroup_rate_options);
+} cgroup_rate_options_map SEC(".maps");
+
+#define NS (1000ULL * 1000ULL * 1000UL)
+
+static inline __attribute__((always_inline)) bool
+cgroup_rate(void *ctx, struct msg_k8s *kube, __u64 time)
+{
+	struct cgroup_rate_options *opt;
+	struct cgroup_rate_key key = {
+		.id = kube->cgrpid,
+	};
+	struct cgroup_rate_value *val;
+	__u64 delta, interval, slide;
+	__u32 zero = 0;
+
+	opt = map_lookup_elem(&cgroup_rate_options_map, &zero);
+	if (!opt)
+		return true;
+
+	interval = opt->interval;
+	if (!interval)
+		return true;
+
+	val = map_lookup_elem(&cgroup_rate_map, &key);
+	if (!val) {
+		struct cgroup_rate_value new_value = {
+			.time = (time / interval) * interval,
+			.curr = 1,
+		};
+
+		map_update_elem(&cgroup_rate_map, &key, &new_value, 0);
```

Comment:
Why is this a per-CPU hash? Is the throttling only per CPU?

Reply:
Okay I guess you want to avoid the race condition that exists even with atomic increments. That makes sense. From a personal perspective, I would have preferred the user space ticker patch to be included in this patch as that answers some of my questions.

Reply:
ugh, sorry about that, I should have included some docs explaining the whole logistics,
there were so many versions of this and I forgot to do that, there's some explanation in https://github.com/cilium/tetragon/pull/2177/commits/363cb62a30710b6f3a925249fbd86ef554df29d0 changelog

basically we compute/monitor the rate per cpu, because I did not find a good way to make atomically,
when cpu rate crosses the limit we send event to user space that triggers throttle start event and triggers user space to periodically check on that cgroup rate, when the rate goes down or the cpu got idle (and all other cpus are ok for the cgroup) we send throttle stop event

we monitor the cgroups per cpu but throttle events are sent for the whole cgroup system wide to prevent getting
many throttle start/stop events from all cpus

having throttle start event means that one or more cpus that execute cgroup code crossed the limit and are throttled

when throttle stop is send that means all previously throttled cpus are now below allowed limit rate

I'll add some documentation